Changes in module 'nf-core/seqkit/split2'
--- modules/nf-core/seqkit/split2/main.nf
+++ modules/nf-core/seqkit/split2/main.nf
@@ -20,14 +20,14 @@
     script:
     def args   = task.ext.args   ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    if(meta.single_end){
+    if(reads instanceof List && reads.size() == 2){
         """
         seqkit \\
             split2 \\
             $args \\
             --threads $task.cpus \\
-            $reads \\
-            --out-dir ${prefix}
+            --read1 ${reads[0]} \\
+            --read2 ${reads[1]}
 
         cat <<-END_VERSIONS > versions.yml
         "${task.process}":
@@ -40,9 +40,7 @@
             split2 \\
             $args \\
             --threads $task.cpus \\
-            --read1 ${reads[0]} \\
-            --read2 ${reads[1]} \\
-            --out-dir ${prefix}
+            $reads
 
         cat <<-END_VERSIONS > versions.yml
         "${task.process}":

************************************************************
