Changes in module 'nf-core/samtools/merge'
--- modules/nf-core/samtools/merge/main.nf
+++ modules/nf-core/samtools/merge/main.nf
@@ -26,6 +26,11 @@
     def args = task.ext.args   ?: ''
     prefix   = task.ext.prefix ?: "${meta.id}"
     def file_type = input_files instanceof List ? input_files[0].getExtension() : input_files.getExtension()
+    if (input_files instanceof List) {
+        sorted_input_files = input_files.toSorted({it.name}).join(' ')
+    } else {
+        sorted_input_files = input_files
+    }
     def reference = fasta ? "--reference ${fasta}" : ""
     """
     samtools \\
@@ -34,7 +39,7 @@
         $args \\
         ${reference} \\
         ${prefix}.${file_type} \\
-        $input_files
+        $sorted_input_files
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
